python/pyspark/streaming/kafka.py

```diff
@@ -81,3 +79,128 @@ def getClassByName(name):
+                                                             decoder, jparam, jtopics)
+        except Py4JError, e:
+            # TODO: use --jar once it also work on driver
+            if not e.message or 'call a package' in e.message:
+                print "No kafka package, please put the assembly jar into classpath:"
+                print " $ bin/spark-submit --driver-class-path external/kafka-assembly/target/" + \
+                      "scala-*/spark-streaming-kafka-assembly-*.jar"
+            raise e
+        ser = PairDeserializer(NoOpSerializer(), NoOpSerializer())
+        stream = DStream(jstream, ssc, ser)
+        return stream.map(lambda (k, v): (keyDecoder(k), valueDecoder(v)))
+
+    @staticmethod
+    def createRDD(sc, brokerList, offsetRanges, kafkaParams={},
+                  keyDecoder=utf8_decoder, valueDecoder=utf8_decoder):
+        """
+        ..note:: experimental
+
+        Create a RDD from Kafka using offset ranges for each topic and partition.
+        :param sc:  SparkContext object
+        :param brokerList: A String representing a list of seed Kafka brokers (hostname:port,...)
+        :param offsetRanges:  list of offsetRange to specify topic:partition [start,
+        end) to consume.
+        :param kafkaParams: Additional params for Kafka
+        :param keyDecoder:  A function used to decode key (default is utf8_decoder)
+        :param valueDecoder:  A function used to decode value (default is utf8_decoder)
+        :return: A RDD object
+        """
+        java_import(sc._jvm, "org.apache.spark.streaming.kafka.KafkaUtils")
+        java_import(sc._jvm, "org.apache.spark.streaming.kafka.OffsetRange")
+
+        kafkaParams.update({"metadata.broker.list": brokerList})
+
+        if not isinstance(offsetRanges, list):
+            raise TypeError("offsetRanges should be list")
+        jparam = MapConverter().convert(kafkaParams, sc._gateway._gateway_client)
+
+        try:
+            array = KafkaUtils._getClassByName(sc._jvm, "[B")
+            decoder = KafkaUtils._getClassByName(sc._jvm, "kafka.serializer.DefaultDecoder")
+            joffsetRanges = sc._gateway.new_array(sc._jvm.OffsetRange, len(offsetRanges))
+            for idx, o in enumerate(offsetRanges):
+                joffsetRanges[idx] = o._joffsetRange(sc)
+            jrdd = sc._jvm.KafkaUtils.createRDD(sc._jsc, array, array, decoder, decoder,
+                                                jparam, joffsetRanges)
+        except Py4JError, e:
+            # TODO: use --jar once it also work on driver
+            if not e.message or 'call a package' in e.message:
+                print "No kafka package, please put the assembly jar into classpath:"
+                print " $ bin/spark-submit --driver-class-path external/kafka-assembly/target/" + \
+                      "scala-*/spark-streaming-kafka-assembly-*.jar"
+            raise e
+        ser = PairDeserializer(NoOpSerializer(), NoOpSerializer())
```

Comment:
There are lots of duplicated code here, could you figure out a way to re-use them? (it's OK to go as this, if it's not easy)